Integrate appointment list

chinzhening opened the pull request:
Closes #85 
This pull request comprises several significant updates aimed at enhancing our appointment management system. Here's a breakdown of the key changes:

-  **Prevent add command from adding overlapping appointments**--Implemented a crucial feature to prevent the addition of overlapping appointments. This ensures the accuracy and coherence of our scheduling system by avoiding conflicting appointments.
-  **Prevent edit command from adding overlapping appointments**--Same as above.
- **Integrate Model component with AddCommand**--Integrated the Model component with the AddCommand feature. This integration streamlines the addition process, enhancing the overall efficiency of our appointment management system.
- **Integrate AppointmentList class with Model component**--Contributed to integrating the AppointmentList class with the Model component. This integration strengthens the backbone of our system, facilitating smoother operations and better organization of appointments. 
- **Provide modifiable and unmodifiable copy support to AppointmentList**--Updated the AppointmentList to support both modifiable and unmodifiable copies. This enhancement provides flexibility in managing appointments while ensuring data integrity and security.

These updates will significantly improve our system's functionality and user experience. I look forward to your feedback and suggestions. Let's merge these changes into the integrate-appointment-list branch to incorporate these improvements seamlessly into our workflow.

**Note that documentation has not been added yet!!!**

chinzhening committed:
Change method name from AppointmentList#hasOverlappingAppointment to
AppointmentList#overlaps

This change will make clearer the intention of the method and also
shorten the method name to improve readability.

chinzhening committed:
Update AppointmentList to support modifiable and unmodifiable copy

chinzhening committed:
Add testing utility class TypicalAppointments to support testing

chinzhening committed:
Integrate AppointmentList class with Model component

ModelManager contains an appointment list which can be serialized into
json form. Preliminary integration testing has also been added.

chinzhening committed:
Integrate Model component with AddCommand.

This is follows the adoption of AppointmentList to keep track of
appointments. Adding a contact with an appointment that clashes with
existing appointments is no longer allowed.

chinzhening committed:
Prevent add command from adding overlapping appointments

chinzhening committed:
Merge branch 'master' into integrate-appointment-list

# Conflicts:
#	src/main/java/seedu/address/logic/commands/AddCommand.java
#	src/main/java/seedu/address/model/Model.java
#	src/main/java/seedu/address/model/ModelManager.java
#	src/test/java/seedu/address/logic/commands/AddCommandTest.java

chinzhening committed:
Update error message for command input validation for appointments

chinzhening committed:
Move appointmentsDoNotOverlap method in AppointmentList to Appointment

chinzhening committed:
Reduce appointment-level operations so AddressBook is still interacted
mainly through person-level operations

chinzhening committed:
Reduce appointment-level operations so AddressBook and ModelManager
are still interacted mainly through person-level operations

chinzhening committed:
Fix json input validation

chinzhening committed:
Fix JsonSerializableAddressBookTest by updating test sample json files

chinzhening committed:
Refactor AddCommand#execute appointments validation

chinzhening committed:
Integrate appointment list with edit command and update Person to store
an AppointmentList instead of Set<Appointment>

Some tweaks to AddCommand is also made

chinzhening committed:
Merge branch 'master' into integrate-appointment-list

# Conflicts:
#	src/main/java/seedu/address/model/util/SampleDataUtil.java

Darylgolden reviewed (changes requested):
I tried out your PR, and when I tried editing an appointment to overlap, no warning shows up in the UI, but
```
INFO: ----------------[USER COMMAND][edit 3 ap/12:00-13:00 SUN]
Exception in thread "JavaFX Application Thread" java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
        at javafx.fxml.FXMLLoader$MethodHandler.invoke(FXMLLoader.java:1857)
        at javafx.fxml.FXMLLoader$ControllerMethodEventHandler.handle(FXMLLoader.java:1724)
        at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:86)
        at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:234)
        at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
        at com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
        at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
        at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
        at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
        at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
        at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
        at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
        at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
        at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:49)
        at javafx.event.Event.fireEvent(Event.java:198)
        at javafx.scene.Node.fireEvent(Node.java:8792)
        at com.sun.javafx.scene.control.behavior.TextFieldBehavior.fire(TextFieldBehavior.java:154)
        at com.sun.javafx.scene.control.behavior.TextInputControlBehavior.lambda$keyMapping$62(TextInputControlBehavior.java:332)
        at com.sun.javafx.scene.control.inputmap.InputMap.handle(InputMap.java:274)
        at com.sun.javafx.event.CompositeEventHandler$NormalEventHandlerRecord.handleBubblingEvent(CompositeEventHandler.java:247)
        at com.sun.javafx.event.CompositeEventHandler.dispatchBubblingEvent(CompositeEventHandler.java:80)
        at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:234)
        at com.sun.javafx.event.EventHandlerManager.dispatchBubblingEvent(EventHandlerManager.java:191)
        at com.sun.javafx.event.CompositeEventDispatcher.dispatchBubblingEvent(CompositeEventDispatcher.java:59)
        at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:58)
        at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
        at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
        at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
        at com.sun.javafx.event.BasicEventDispatcher.dispatchEvent(BasicEventDispatcher.java:56)
        at com.sun.javafx.event.EventDispatchChainImpl.dispatchEvent(EventDispatchChainImpl.java:114)
        at com.sun.javafx.event.EventUtil.fireEventImpl(EventUtil.java:74)
        at com.sun.javafx.event.EventUtil.fireEvent(EventUtil.java:54)
        at javafx.event.Event.fireEvent(Event.java:198)
        at javafx.scene.Scene$KeyHandler.process(Scene.java:4105)
        at javafx.scene.Scene.processKeyEvent(Scene.java:2156)
        at javafx.scene.Scene$ScenePeerListener.keyEvent(Scene.java:2630)
        at com.sun.javafx.tk.quantum.GlassViewEventHandler$KeyEventNotification.run(GlassViewEventHandler.java:218)
        at com.sun.javafx.tk.quantum.GlassViewEventHandler$KeyEventNotification.run(GlassViewEventHandler.java:150)
        at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
        at com.sun.javafx.tk.quantum.GlassViewEventHandler.lambda$handleKeyEvent$1(GlassViewEventHandler.java:250)
        at com.sun.javafx.tk.quantum.QuantumToolkit.runWithoutRenderLock(QuantumToolkit.java:424)
        at com.sun.javafx.tk.quantum.GlassViewEventHandler.handleKeyEvent(GlassViewEventHandler.java:249)
        at com.sun.glass.ui.View.handleKeyEvent(View.java:548)
        at com.sun.glass.ui.View.notifyKey(View.java:972)
        at com.sun.glass.ui.win.WinApplication._runLoop(Native Method)
        at com.sun.glass.ui.win.WinApplication.lambda$runLoop$3(WinApplication.java:184)
        at java.base/java.lang.Thread.run(Thread.java:833)
Caused by: java.lang.reflect.InvocationTargetException
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at com.sun.javafx.reflect.Trampoline.invoke(MethodUtil.java:77)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:568)
        at com.sun.javafx.reflect.MethodUtil.invoke(MethodUtil.java:275)
        at com.sun.javafx.fxml.MethodHelper.invoke(MethodHelper.java:84)
        at javafx.fxml.FXMLLoader$MethodHandler.invoke(FXMLLoader.java:1854)
        ... 46 more
Caused by: seedu.address.model.appointment.exceptions.OverlappingAppointmentException: Operation would result in overlapping appointments
        at seedu.address.model.appointment.AppointmentList.add(AppointmentList.java:41)
        at seedu.address.model.AddressBook.setPerson(AddressBook.java:118)
        at seedu.address.model.ModelManager.setPerson(ModelManager.java:126)
        at seedu.address.logic.commands.EditCommand.execute(EditCommand.java:99)
        at seedu.address.logic.LogicManager.execute(LogicManager.java:51)
        at seedu.address.ui.MainWindow.executeCommand(MainWindow.java:176)
        at seedu.address.ui.CommandBox.handleCommandEntered(CommandBox.java:45)
        ... 58 more
```
shows in the command output instead.
When I do it again, it just allows it to happen.
![image](https://github.com/AY2324S2-CS2103-F09-3/tp/assets/9102529/43c4e319-85ac-43b7-b089-5a4adc263989)

chinzhening committed:
Fix bug with edit command and refactor AppointmentList class to
DisjointAppointmentList (for AddressbBook) and AppointmentList

Darylgolden commented:
@biinnnggggg Can you add support for handling the case of loading from storage if that storage has overlapping appointments?
Technically it shouldn't happen if the user is only using 1.3, but it's better to be defensive.

chinzhening committed:
Remove setAppointment functionality from Model

Reason: model's appointment list should only be edited through
person-level interactions.

chinzhening committed:
Update Json loading to be more defensive against overlapping
appointments

chinzhening committed:
Remove redundant method from ModelStub

chinzhening committed:
Update ModelClassDiagram

chinzhening committed:
Add details to Model component and Implementation section of adding appt

Darylgolden reviewed (approved):
This works for me, thanks!
